### DISCUSSION TITLE
il9341 with xpt2046

### DISCUSSION THREAD

**Discussion by fifa255** [arendst/Tasmota#21227] [Q&A] [answered]
I am trying to get Haspmota working on my esp32 connected to a il9341 with xpt2046 module.  I have previously used the same setup with openhasp and now trying out haspmota.  The same setup I'm using worked with both display and touch with openhasp

I am using the display module referenced here on the openhasp page:  https://www.openhasp.com/0.6.3/displays/MSPxxxx/
the pin connections for esp32 and display module are as described below

![image](https://github.com/arendst/Tasmota/assets/24857898/9c207e56-9b4b-4d62-9358-bd7390198e03)

Module configuration in tasmota:
![image](https://github.com/arendst/Tasmota/assets/24857898/a20625a6-cbfb-4376-bc57-340d14fde1b4)

I am using the [ILI9341_xpt_display.ini](https://github.com/arendst/Tasmota/blob/development/tasmota/displaydesc/ILI9341_xpt_display.ini)

with the following added below
:TS,*
:M

I have the display working but I am unable to get touch working. I have uploaded the [DisplayCalibrate.tapp](https://raw.githubusercontent.com/arendst/Tasmota/development/tasmota/berry/modules/DisplayCalibrate.tapp) and the calibration process starts and asks me to touch the top left cross. But my display is not registering any touch events and does not respond.

Can you please guide me in what maybe happening here? Are the pin connections wrong or the display.ini that I am using wrong?  Any help would be appreciated. Thanks in advance.

**Comment by s-hadinger** [comment, marked as answer]
We have migrated to Universal Touch driver (uTouch), see https://tasmota.github.io/docs/Universal-Display-Driver/#migrating-to-utouch

You need to remove `:TS` and replace with:

```
:UTI,XPT2046,S1,*,-1,*
CPR 00
RT
:UTT
XPT 300
RT
:UTX
MV 0 2
RT
:UTY
MV 2 2
RT
#
```

You also need to set:
- GPIO 17 to `TS SPI CS`

Only after this you need to run DisplayCalibrate

**Reply by fifa255** [reply]
Thank you. looks like I needed to install the development branch for the utouch. I made the above changes and added 

:M 

to the end of the display.ini and ran displaycalibrate and it worked. Thank you for your help.

**Comment by fifa255** [comment]
So after getting the display calibration completed I am having trouble reacting to a single test button

I have the following loaded into the file system in addition to the display.ini. with the added  :M,2091,2120,2065,2047 automatically generated after the calibration test.

autoexec.be

import haspmota
haspmota.start()

pages.jsonl

{"page":0,"comment":"---------- Upper stat line ----------"}

{"id":11,"obj":"label","x":0,"y":0,"w":320,"pad_right":90,"h":22,"bg_color":"#D00000","bg_opa":255,"radius":0,"border_side":0,"text":"Tasmota","text_font":"montserrat-20"}

{"id":15,"obj":"lv_wifi_arcs","x":291,"y":0,"w":29,"h":22,"radius":0,"border_side":0,"bg_color":"#000000","line_color":"#FFFFFF"}
{"id":16,"obj":"lv_clock","x":232,"y":3,"w":55,"h":16,"radius":0,"border_side":0}

{"page":1,"comment":"---------- Page 1 ----------"}
{"id":0,"bg_color":"#0000A0","bg_grad_color":"#000000","bg_grad_dir":1,"text_color":"#FFFFFF"}
{"id":1,"obj":"btn","x":100,"y":60,"w":80,"h":30,"bg_color":"#4f4f4f","text_color":"#FFFF44","radius":10,"text":"Enabled","border_width":3,"border_side":15,"border_color":"#FFFF44","border_opa":210}

{"berry_run":"tasmota.add_rule('hasp#p1b1#event=up', def () print('Button p1b1 pressed') end)"}

I expect to see a message "Button p1b1 pressed" displayed in the berry console when I click on the button but nothing happens. 

This is the startup message from tasmota console 

00:00:00.001 HDW: ESP32-D0WDQ6 v1.0 
00:00:00.040 UFS: FlashFS mounted with 296 kB free
00:00:00.057 CFG: Loaded from File, Count 38
00:00:00.067 QPC: Count 1
00:00:00.069 SPI: Hardware using GPIO18(CLK), GPIO23(MOSI) and GPIO19(MISO)
00:00:00.122 BRY: Berry initialized, RAM used 5321 bytes
00:00:00.855 UT : XPT2046
00:00:00.855 DSP: Configured display 'ILI9341'
00:00:00.865 Project tasmota - Tasmota Version 13.4.1.2(805ad32-lvgl-haspmota)-3_0_0(2024-04-21T21:12:20)
00:00:00.867 LVG: Allocating buffer1 28 KB in main memory (flushlines 40)
00:00:00.882 LVG: LVGL initialized
00:00:01.252 BRY: Successfully loaded 'autoexec.be'

What am I doing wrong?  Where does the code for button event reactions go?  Thank you for the help.

**Reply by sfromis** [reply]
You should see the button visually react to being pushed, if the calibration is spot on.

To see if you get any touch events, including outside of the button, you could try something like this for debugging:
`{"berry_run":"tasmota.add_rule('hasp#?#?', /val,trig,data -> print(val, trig, data))"}`

Running with log level 4 for "more debug", the console messages are also pretty informative about what happens with touch.

**Comment by fifa255** [comment]
Setting the loglevel to 4 on seriallog does report touch events and their coordinates and raw coordinates. the calibration data obtained from displaycalibrate does not seem to be accurate

:M,2054,2022,2054,2020

This calibration makes touches on the screen mapped to x=0 or 240 and y= 0 or 320 with moving just a few pixels. I can see the raw x and y values changing with more precision but the x and y values seem to be flipped. My display is oriented vertically when set to displayrotate 0. THe display is 240 w and 320 h in this orientation.  The mapping of the raw x/y values to screen x/y values seem to off and flipped. 

Here's what I get now if the click the 4 corners of the screen after running displaycalibrate

Clicking on the top left corner 
 TS : touched  x=0 y=320 gest=0x00 (raw x=3729 y=443)

Top right
 TS : touched  x=0 y=0 gest=0x00 (raw x=3693 y=3659)

Bottom Left 
TS : touched  x=240 y=320 gest=0x00 (raw x=436 y=491)

Bottom Right
 TS : touched  x=240 y=0 gest=0x00 (raw x=450 y=3666)

How can I correct this?  appreciate your help.

**Reply by s-hadinger** [reply]
Do you see the button visually react when pressing on it? (zoom effect)

**Reply by sfromis** [reply]
Those calibration numbers in `:M` look bad, a failed calibration. With the raw values you show, I'd expect that the 4 numbers should be around 400 and around 3700.

**Reply by fifa255** [reply]
No zoom effect on button press because I don't think the touch is being registered at the correct position.  I have tried the display calibration app several times and get the same or similar valleys for the M line.

I have tried to manually get the calibration numbers by looking at the raw x and y values at the four corners and place them in display.ini but the x and y axes seem to be transposed.

I found the following info from the documentation fo descriptor file.

M,X1,X2,Y1,Y2 Defines an optional mapping for touch controllers (always needed on resistive touch) X1 = display left margin
X2 = display right margin
Y1 = display upper margin
Y2 = display lower margin


If I use the raw values off the four corners from the earlier post I get 

3729,450,443,3666.

I have tried similar values for the :M line and can get a better result but the axes seems flipped.

**Reply by sfromis** [reply]
When it comes to flipping, you may have to adjust the rotation settings in `display.ini`, where you can flip and swap x and y.
https://tasmota.github.io/docs/Displays/#descriptor-file
Look for "rotation pseudo opcode" and the text around that point.

**Comment by fifa255** [comment]
I am quite frustrated trying to get this display and touch to work. I have tried to play around with the pseudo opcodes to try to flip and swap x and y but not getting the result I want.   The below orientation and x,y coordinates are what I am trying to achieve for touch.

![image](https://github.com/arendst/Tasmota/assets/24857898/4d88725b-feaa-4f7d-b4ef-420177b0b787)

With the default rotation opcode and :M line

:R,36
:0,48,00,00,00
:1,28,00,00,01
:2,88,00,00,02
:3,E8,00,00,03

:M,3764,376,373,3699

I get the following result for x,y coordinate for the four corners

![image](https://github.com/arendst/Tasmota/assets/24857898/63bb9bfe-2cdd-4fe4-910f-2197fb01353d)

So it seem I need to swap x and y axes. 

The tamota documentation for rotation opcodes mentions bit 7 for swaping x and y which I assumes means in binary 0100 0000 which is 20 in hex.  So making the change to the rotation opcode above results in with last value changed to 20 for rotation 0

:R,36
:0,48,00,00,20
:1,28,00,00,01
:2,88,00,00,02
:3,E8,00,00,03

:M,3764,376,373,3699

But this gives me the exact same values as above for the four corners. Am I doing something wrong? Please help as I am at my wits end trying to get the touch to work accurately. Thank you

**Reply by fifa255** [reply]
sorry I meant 40 in hex rather than 20

**Reply by tplet** [reply]
Thank you for your message, which helped me move forward.

I'm having a similar problem but in landscape mode (ST7789 screen, 240x320 with XPT2046 touch driver).

In portrait mode, the configuration of parameters :0 and :2 works fine, but in :1 and :3, there is a shift. X and Y seem to be correctly configured, but X goes from 0 (left) to 240 (right), and Y from 0 (top) to 340 (bottom)... I'm in landscape. X should go to max 340 and Y to max 240.

Could you give me an idea?

Here is my display.ini : 
```suggestion
:H,ST7789,240,320,16,SPI,1,*,*,*,*,*,*,*,40
:S,2,1,3,0,80,30
:I
01,A0
11,A0
3A,81,55
36,81,00
21,80
13,80
29,A0
:o,28
:O,29
:R,36
:0,C0,00,00,84
:1,A0,00,00,00
:2,00,00,00,85
:3,60,00,00,03
:i,20,21
:A,2A,2B,2C
:B,20,4
:M,3719,223,3892,253
:r,1
:UTI,XPT2046,S1,*,-1,*
CPR 00
RT
:UTT
XPT 300
RT
:UTX
MV 0 2
;DBG 1
RT
:UTY
MV 2 2
RT
#
```

**Comment by joba-1** [comment]
Bit numbering starts at 0 for 2^0
So bit 7 = 2^7 = 128 dec or 80 hex

**Comment by fifa255** [comment]
Oh finally got touch working accurately!!! Thank you.
